src/main/scala/Move.scala

```diff
@@ -39,24 +39,26 @@ case class Move(
             else h1.halfMoveClock + 1
         )
 
-        val halfCastlingRights: Castles =
-          if captures then h1.castles & ~dest.bitboard
-          else h1.castles
-
-        val castleRights: Castles =
-          if (piece is Rook) && (orig.bitboard & h2.unmovedRooks).nonEmpty then
-            halfCastlingRights & ~orig.bitboard
-          else if piece.is(King) then halfCastlingRights & Bitboard.rank(piece.color.lastRank)
-          else halfCastlingRights
-
-        var unmovedRooks: UnmovedRooks =
-          if captures then h2.unmovedRooks & ~dest.bitboard
-          else h2.unmovedRooks
-
-        unmovedRooks =
-          if piece is Rook then unmovedRooks & ~orig.bitboard
-          else if piece is King then unmovedRooks & Bitboard.rank(piece.color.lastRank)
-          else unmovedRooks
+        var castleRights: Castles      = h2.castles
+        var unmovedRooks: UnmovedRooks = h2.unmovedRooks
+
+        // If the rook is captured
+        // update castling rights and unmoved rooks bitboard.
+        if captures && (h2.unmovedRooks & dest.bitboard).nonEmpty then
+          unmovedRooks = unmovedRooks & ~dest.bitboard
+          castleRights = castleRights & ~dest.bitboard
```

Comment:
This passes all the tests, but not actually correct, because castleRights and unmmovedRooks are not the same squares.